File: src/main/java/tweetzor/core/Tweet.java
package tweetzor.core;

import twitter4j.Status;

public class Tweet {

  private long id;
  private String author;
  private String text;
  private long retweetCount;
  private Location location;

  public Tweet(long id, String author, String text, long retweetCount, Location location) {
    this.id = id;
    this.author = author;
    this.text = text;
    this.retweetCount = retweetCount;
    this.location = location;
  }

  public Tweet(Status status) {
    this.id = status.getId();
    this.author = status.getUser().getName();
    this.text = status.getText();
    this.retweetCount = status.getRetweetCount();
    this.location = status.getGeoLocation() == null ? null : new Location(status.getGeoLocation());
  }

  public long getId() {
    return id;
  }

  public void setId(long id) {
    this.id = id;
  }

  public String getAuthor() {
    return author;
  }

  public void setAuthor(String author) {
    this.author = author;
  }

  public String getText() {
    return text;
  }

  public void setText(String text) {
    this.text = text;
  }

  public long getRetweetCount() {
    return retweetCount;
  }

  public void setRetweetCount(long retweetCount) {
    this.retweetCount = retweetCount;
  }

  public Location getLocation() {
    return location;
  }

  public void setLocation(Location location) {
    this.location = location;
  }
  
}

File: src/main/scala/tweetzor2/core/Tweet.scala
package tweetzor2.core

import twitter4j.Status

// The `case class` keywords generate a class with a bunch of predefined convenience functionality:
// 
// * primary constructor;
// * accessors for all fields;
// * mutators for all mutable fields;
// * `copy` method for functional updates;
// * working `equals` and `hashCode` methods;
// * sensible `toString` method;
// * companion object with `apply` and `unapply` methods.
case class Tweet(
  val id: Long,
  val author: String,
  val text: String,
  val retweetCount: Long,
  val location: Option[Location]
)

// The `object` keyword defines a singleton object.
// In this case, we are defining a *companion object* and factory method for the `Tweet` class.
//
// Singleton objects provide a replacement for static fields and methods.
// They have the advantage of being fully fledged objects that can extend a class.
//
// Companion classes and objects have special scoping privileges 
// that make them well suited for factory-like patterns.
object Tweet {
  def apply(status: Status): Tweet = {
    Tweet(
      id = status.getId,
      author = status.getUser.getName,
      text = status.getText,
      retweetCount = status.getRetweetCount,
      location = Option(status.getGeoLocation).map(Location.apply _)
    )
  }
}
